File: scripts/spike-marks.cs
#:package YDotNet@0.6.0
#:package YDotNet.Native@0.6.0

// ═══════════════════════════════════════════════════════════════════════════
//  spike: YDotNet เขียน "mark" (ตัวหนา/ลิงก์) ที่ yjs 13.x อ่านออกได้จริงไหม
//
//      dotnet run scripts/spike-marks.cs | node scripts/spike-marks.mjs
//
//  ⚠️ นี่คือจุดยกเลิกของงานใส่ตัวหนา/เอียง/ลิงก์
//
//     API ฝั่ง C# มีครบ (XmlText.Format + Input.Object — reflect ดูแล้ว) และ
//     ฝั่ง JS พิสูจน์แล้วว่า mark ที่ถูกต้องผ่าน node.check() ได้ แต่ "สะพาน"
//     ระหว่างสองฝั่ง — ว่า ContentFormat ที่ yrs เขียนออกมา decode ได้ตรงกับที่
//     yjs 13.x คาดหวังหรือไม่ — ยังไม่มีใครพิสูจน์
//
//     ต้องรู้คำตอบก่อนเขียนโค้ดจริง เพราะความล้มเหลวแบบครึ่ง ๆ กลาง ๆ ของเรื่องนี้
//     คือ "ข้อความถูกลบเงียบ ๆ โดยที่ node.check() ยังผ่าน" (ทดลองยืนยันแล้ว)
//     ไม่ใช่แค่แสดงผลไม่สวย
//
//  ⚠️ offset เป็นหน่วย UTF-16 code unit — DocOptions.Encoding ปล่อยเป็นค่าเริ่มต้น
//     (Utf16) ซึ่งตรงกับทั้ง C# string.Length และ yjs ห้ามไปตั้งเป็น Utf8
// ═══════════════════════════════════════════════════════════════════════════

using YDotNet.Document;
using YDotNet.Document.Cells;
using YDotNet.Document.Options;

const string Plain = "ปกติ";
const string Bold = "หนา";
const string Linked = "ลิงก์";
const string Url = "https://example.test/ก";

var doc = new Doc(new DocOptions { Id = 4242 });
var fragment = doc.XmlFragment("blocknote");

byte[] before;
using (var read = doc.ReadTransaction()) before = read.StateVectorV1();

using (var write = doc.WriteTransaction(origin: null!))
{
    var group = fragment.InsertElement(write, 0, "blockGroup");
    var container = group.InsertElement(write, 0, "blockContainer");
    container.InsertAttribute(write, "id", Guid.CreateVersion7().ToString());

    var paragraph = container.InsertElement(write, 0, "paragraph");
    var text = paragraph.InsertText(write, 0);

    // ⚠️ Insert ครั้งเดียวแล้วค่อย Format เป็นช่วง ๆ
    //    Insert ที่ attributes เป็น null "สืบทอดรูปแบบจากตำแหน่งนั้น" ซึ่งทำให้
    //    ข้อความที่เขียนต่อจากช่วงตัวหนากลายเป็นตัวหนาไปด้วยโดยไม่ตั้งใจ
    text.Insert(write, 0, Plain + Bold + Linked, null!);

    using var bold = Input.Object(new Dictionary<string, Input>
    {
        // ค่าของ mark คือ "attrs object" ของมัน — ตัวที่ไม่มี attr ใช้ object ว่าง
        ["bold"] = Input.Object(new Dictionary<string, Input>()),
    });
    text.Format(write, (uint)Plain.Length, (uint)Bold.Length, bold);

    using var link = Input.Object(new Dictionary<string, Input>
    {
        ["link"] = Input.Object(new Dictionary<string, Input> { ["href"] = Input.String(Url) }),
    });
    text.Format(write, (uint)(Plain.Length + Bold.Length), (uint)Linked.Length, link);

    write.Commit();
}

using var diff = doc.ReadTransaction();
Console.WriteLine(Convert.ToBase64String(diff.StateDiffV1(before)));

File: scripts/spike-marks.mjs
#!/usr/bin/env node
// ═══════════════════════════════════════════════════════════════════════════
//  อีกครึ่งของ spike — รับ base64 จาก stdout ของ spike-marks.cs
//
//      dotnet run scripts/spike-marks.cs | node scripts/spike-marks.mjs
//
//  ตอบคำถามเดียว: mark ที่ YDotNet เขียน yjs 13.x + schema จริงของ BlockNote
//  อ่านออกไหม และ "ข้อความยังอยู่ครบ" ไหม
//
//  ⚠️ ข้อหลังสำคัญกว่าข้อแรก — ทดลองแล้วว่าถ้าชื่อ mark ไม่ตรงกับ schema
//     ข้อความจะถูกลบทิ้งโดยที่ node.check() ยังผ่าน
// ═══════════════════════════════════════════════════════════════════════════
import { readFile } from 'node:fs/promises'

async function importEsm(name) {
  const dir = new URL(`../web/node_modules/${name}/`, import.meta.url)
  const pkg = JSON.parse(await readFile(new URL('package.json', dir), 'utf8'))
  return import(new URL(pkg.exports?.['.']?.import ?? pkg.module ?? pkg.main, dir).href)
}

const Y = await importEsm('yjs')
const { BlockNoteEditor } = await importEsm('@blocknote/core')
const { yXmlFragmentToProseMirrorRootNode } = await importEsm('y-prosemirror')

const C = { red: '\x1b[31m', green: '\x1b[32m', dim: '\x1b[2m', bold: '\x1b[1m', off: '\x1b[0m' }

let failed = 0
function check(label, ok, detail = '') {
  if (ok) console.log(`  ${C.green}✓${C.off} ${label}`)
  else {
    failed++
    console.log(`  ${C.red}✗${C.off} ${label}${detail ? `\n      ${C.dim}${detail}${C.off}` : ''}`)
  }
}

// fs/promises.readFile ไม่รับ fd 0 — อ่านจาก stream แทน
const chunks = []
for await (const chunk of process.stdin) chunks.push(chunk)
const base64 = Buffer.concat(chunks).toString('utf8').trim().split('\n').at(-1)
const update = Uint8Array.from(Buffer.from(base64, 'base64'))

const doc = new Y.Doc()
Y.applyUpdate(doc, update)

const schema = BlockNoteEditor.create().pmSchema
const node = yXmlFragmentToProseMirrorRootNode(doc.getXmlFragment('blocknote'), schema)

console.log(`${C.bold}spike: mark จาก YDotNet ข้ามไปฝั่ง yjs ได้ไหม${C.off}\n`)

try {
  node.check()
  check('node.check() ผ่าน', true)
} catch (error) {
  check('node.check() ผ่าน', false, error.message)
}

const expected = 'ปกติหนาลิงก์'
check('ข้อความครบทุกไบต์ (ไม่ถูกลบเงียบ ๆ)',
  node.textContent === expected,
  `${JSON.stringify(node.textContent)} ≠ ${JSON.stringify(expected)}`)

const runs = []
node.descendants((child) => {
  if (child.isText) runs.push({ text: child.text, marks: child.marks.map((m) => m.type.name) })
})
console.log(`  ${C.dim}runs: ${JSON.stringify(runs)}${C.off}`)

check('"ปกติ" ไม่มี mark',
  runs.find((r) => r.text === 'ปกติ')?.marks.length === 0)
check('"หนา" มี mark bold',
  runs.find((r) => r.text === 'หนา')?.marks.includes('bold') === true)

const linkRun = runs.find((r) => r.text === 'ลิงก์')
check('"ลิงก์" มี mark link', linkRun?.marks.includes('link') === true)

let href = null
node.descendants((child) => {
  if (child.isText) {
    const mark = child.marks.find((m) => m.type.name === 'link')
    if (mark) href = mark.attrs.href
  }
})
check('link เก็บ href มาด้วย', href === 'https://example.test/ก', String(href))

console.log(failed === 0
  ? `\n${C.green}${C.bold}ทำได้ — ไปต่อ S4 ได้${C.off}\n`
  : `\n${C.red}${C.bold}ไม่ผ่าน ${failed} ข้อ — ยกเลิก mark แล้วลดรูปแทน${C.off}\n`)

process.exit(failed > 0 ? 1 : 0)
